### ISSUE TITLE
Diacritics compositions / dotted circle ligatures

### ISSUE THREAD

**Issue by dy** [google/fonts#2974]
To be able to support diacritics compositions, google fonts need to include support for 
`dotted circle + *` ligatures (type 4 in GSUB)
![image](https://user-images.githubusercontent.com/300067/105726030-a13d1500-5ef7-11eb-8be6-a354b53941ba.png)
![image](https://user-images.githubusercontent.com/300067/105726057-a7cb8c80-5ef7-11eb-9a8e-55130f598ab2.png)
Afaik there's no that feature yet.
Ref https://docs.microsoft.com/en-us/typography/opentype/spec/gsub, https://docs.microsoft.com/en-us/typography/script-development/arabic

_Originally posted by @dy in https://github.com/google/fonts/issues/1808#issuecomment-766894187_

**Comment by tiroj** [comment]
A font really shouldn’t require ligatures of the dotted circle (U+25CC) as a generic mark holder with either individual marks or combinations of marks, since this can be handled using mark GPOS. The shaping engine should automatically produce the result shown in the image if the font includes handling for the two-mark sequence—either with a mark ligature or using mkmk positioning—and mark positioning on the U+25CC base.

Not all shaping engines automatically insert a dotted circle base glyph before unsupported marks (I think Apple’s CoreText engine does not), so results from fonts will vary depending on platform.

**Comment by dy** [comment]
You're right. Combining marks can be added automatically-ish via feature file:

```lookup diacritics {
  markClass [gravecomb acutecomb] <anchor 160 512> @TOP_MARKS;

  # ш м т
  pos base [\uni0448 \uni043C \uni0442] <anchor 500 512> mark @TOP_MARKS;
  # ...rest of glyphs
} diacritics;

feature mark {
  lookup diacritics;
} mark;

# compile as 
# makeotf -ff diacritics.fea -f source.ttf -o result.otf
```

**Comment by dy** [comment]
Temporarily there's a fix for Lora/Noto fonts https://github.com/google/fonts/issues/2974
